src/Compilers/Core/Portable/Syntax/AbstractWarningStateMap.cs

```diff
@@ -50,7 +50,7 @@ public ReportDiagnostic GetWarningState(string id, int position)
         private WarningStateMapEntry GetEntryAtOrBeforePosition(int position)
         {
             Debug.Assert(_warningStateMapEntries != null && _warningStateMapEntries.Length > 0);
-            int r = Array.BinarySearch(_warningStateMapEntries, new WarningStateMapEntry(position));
+            int r = Array.BinarySearch(_warningStateMapEntries, new WarningStateMapEntry(position), new WarningStateMapEntryComparer ());
             return _warningStateMapEntries[r >= 0 ? r : ((~r) - 1)];
```

Comment:
This should be WarningStateMapEntryComparer.Instance, once you've made it a singleton.